File: files/300-399/375/CH-375.py
import pandas as pd

path = "300-399/375/CH-375 Table Transformation.xlsx"
input = pd.read_excel(path, usecols="B:C", skiprows=2, nrows=9)
test = pd.read_excel(path, usecols="E:G", skiprows=2, nrows=5)

input_matrix = input.values
input_array = input_matrix.flatten()
input_array = input_array[~pd.isnull(input_array)]
cell_types = len(set([type(cell) for cell in input_array]))
input_array = input_array.reshape(-1, cell_types)
input_df = pd.DataFrame(input_array, columns=["Date", "Product", "Sale"])

print((input_df == test).all().all())

File: files/300-399/375/CH-375.R
library(tidyverse)
library(readxl)

path <- "300-399/375/CH-375 Table Transformation.xlsx"
input <- read_excel(
  path,
  range = "B4:C11",
  col_types = c("text", "text"),
  col_names = F
) %>%
  as.matrix()
test <- read_excel(path, range = "E3:G8")

result = input %>%
  matrix(., ncol = 1, byrow = TRUE) %>%
  na.omit() %>%
  as.data.frame() %>%
  mutate(group = cumsum(str_detect(V1, "^[0-9]{5}"))) %>%
  mutate(nr = row_number(), .by = group) %>%
  pivot_wider(names_from = nr, values_from = V1) %>%
  select(Date = `1`, Product = `3`, Sale = `2`) %>%
  mutate(
    Date = janitor::excel_numeric_to_date(as.numeric(Date)) %>% as.POSIXct(),
    Sale = as.numeric(Sale)
  ) %>%
  arrange(Date, desc(Product), desc(Sale))

all.equal(result, test, check.attributes = FALSE)
# different Product mapping.
